flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/SlotSharingExecutionSlotAllocatorTest.java

```diff
@@ -0,0 +1,575 @@
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.scheduler;
+
+import org.apache.flink.runtime.clusterframework.types.AllocationID;
+import org.apache.flink.runtime.clusterframework.types.ResourceProfile;
+import org.apache.flink.runtime.clusterframework.types.SlotProfile;
+import org.apache.flink.runtime.executiongraph.utils.SimpleAckingTaskManagerGateway;
+import org.apache.flink.runtime.instance.SimpleSlotContext;
+import org.apache.flink.runtime.jobgraph.JobVertexID;
+import org.apache.flink.runtime.jobmaster.LogicalSlot;
+import org.apache.flink.runtime.jobmaster.SlotRequestId;
+import org.apache.flink.runtime.jobmaster.TestingPayload;
+import org.apache.flink.runtime.jobmaster.slotpool.PhysicalSlot;
+import org.apache.flink.runtime.jobmaster.slotpool.PhysicalSlotProvider;
+import org.apache.flink.runtime.jobmaster.slotpool.PhysicalSlotRequest;
+import org.apache.flink.runtime.scheduler.SharedSlotProfileRetriever.SharedSlotProfileRetrieverFactory;
+import org.apache.flink.runtime.scheduler.strategy.ExecutionVertexID;
+import org.apache.flink.runtime.taskmanager.LocalTaskManagerLocation;
+import org.apache.flink.util.FlinkRuntimeException;
+import org.apache.flink.util.Preconditions;
+
+import org.junit.Test;
+import org.powermock.core.IdentityHashSet;
+
+import javax.annotation.Nullable;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.IdentityHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.CancellationException;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.function.BiConsumer;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+import static org.hamcrest.CoreMatchers.is;
+import static org.hamcrest.Matchers.containsInAnyOrder;
+import static org.hamcrest.Matchers.hasSize;
+import static org.hamcrest.Matchers.instanceOf;
+import static org.hamcrest.Matchers.notNullValue;
+import static org.junit.Assert.assertThat;
+import static org.junit.Assert.fail;
+
+/**
+ * Test suite for {@link SlotSharingExecutionSlotAllocator}.
+ */
+public class SlotSharingExecutionSlotAllocatorTest {
+	@Test
+	public void testSlotProfileRequestAskedBulkAndGroup() {
+		AllocationContext context = AllocationContext.newBuilder().setGroups(2).build();
```

Comment:
I feel that using indices here and there makes it a bit hard for reading and reasoning.
How about let the test case decide the grouping in a more direct way.

Below is an example showing a way that I feel more readable. I also implies some implementation changes on `AllocationContext`.

```
public class SlotSharingExecutionSlotAllocatorTest {

	private static final ExecutionVertexID EV1 = createRandomExecutionVertexID();
	private static final ExecutionVertexID EV2 = createRandomExecutionVertexID();
	private static final ExecutionVertexID EV3 = createRandomExecutionVertexID();
	private static final ExecutionVertexID EV4 = createRandomExecutionVertexID();

	@Test
	public void testNewAllocatePhysicalSlotForSharedSlot() {
		List<List<ExecutionVertexID>> vertices = Arrays.asList(
			Arrays.asList(EV1, EV2),
			Arrays.asList(EV3, EV4));
		AllocationContext context = AllocationContext.newBuilder().setVertices(vertices).build();

		List<SlotExecutionVertexAssignment> executionVertexAssignments = context.allocateSlotsFor(EV1, EV2, EV3, EV4);
		Collection<ExecutionVertexID> assignIds = getAssignIds(executionVertexAssignments);

		assertThat(assignIds, containsInAnyOrder(EV1, EV2, EV3, EV4));
		assertThat(context.getSlotProvider().getRequests().keySet(), hasSize(2));
	}

	@Test
	public void testSlotProfileRequestAskedBulkAndGroup() {
		List<ExecutionVertexID> vertexGroup1 = Arrays.asList(EV1, EV2);
		List<List<ExecutionVertexID>> vertices = Arrays.asList(vertexGroup1);
		AllocationContext context = AllocationContext.newBuilder().setVertices(vertices).build();
		ExecutionSlotSharingGroup executionSlotSharingGroup =
			context.getSlotSharingStrategy().getExecutionSlotSharingGroup(EV1);

		context.allocateSlotsFor(EV1, EV2);

		Set<ExecutionVertexID> ids = new HashSet<>(vertexGroup1);
		assertThat(context.getSlotProfileRetrieverFactory().getAskedBulks(), containsInAnyOrder(ids));
		assertThat(context.getSlotProfileRetrieverFactory().getAskedGroups(), containsInAnyOrder(executionSlotSharingGroup));
	}
...
```